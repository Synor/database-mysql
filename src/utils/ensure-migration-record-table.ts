import { performance } from 'perf_hooks'

type QueryStore = import('../queries').QueryStore

export const ensureMigrationRecordTable = async (
  queryStore: QueryStore,
  baseVersion: string
): Promise<void> => {
  const existingColumnNames = await queryStore.getMigrationTableColumnNames()

  const migrationTableExists = Boolean(existingColumnNames.length)

  const startTime = performance.now()

  if (!migrationTableExists) {
    await queryStore.createMigrationTable()
  }

  for (const [columnName, addColumn] of Object.entries(queryStore.addColumn)) {
    if (!existingColumnNames.includes(columnName)) {
      await addColumn()
    }
  }

  const endTime = performance.now()

  if (!migrationTableExists) {
    await queryStore.addRecord({
      version: baseVersion,
      type: 'do',
      title: 'Base Migration',
      hash: '',
      appliedAt: new Date(),
      appliedBy: 'Synor',
      executionTime: endTime - startTime,
      dirty: false
    })
  }
}
