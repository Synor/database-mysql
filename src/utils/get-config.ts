import { SynorError } from '@synor/core'
import { ConnectionString } from 'connection-string'
import { readFileSync } from 'fs'
import { resolve as resolvePath } from 'path'

type ConnectionConfig = import('mysql').ConnectionConfig
type TLSConnectionOptions = import('tls').ConnectionOptions

type SSLConfig = Pick<
  TLSConnectionOptions,
  'ca' | 'cert' | 'ciphers' | 'key' | 'passphrase' | 'rejectUnauthorized'
>

type SSLParams = Pick<
  SSLConfig,
  'ciphers' | 'passphrase' | 'rejectUnauthorized'
> & {
  ca?: string
  key?: string
  cert?: string
}

type MySQLDatabaseConfig = Required<Pick<ConnectionConfig, 'database'>> &
  Pick<
    ConnectionConfig,
    'host' | 'port' | 'user' | 'password' | 'multipleStatements' | 'ssl'
  >

type MySQLEngineConfig = {
  migrationRecordTable: string
}

export function getConfig(
  uri: string
): {
  databaseConfig: MySQLDatabaseConfig
  engineConfig: MySQLEngineConfig
} {
  try {
    const {
      protocol,
      hostname: host,
      port,
      user,
      password,
      path,
      params = {}
    } = new ConnectionString(uri, {
      params: {
        synor_migration_record_table: 'synor_migration_record'
      }
    })

    if (!protocol) {
      throw new Error(`[URI] missing: protocol!`)
    }

    if (protocol !== 'mysql') {
      throw new Error(`[URI] unsupported: protocol(${protocol})!`)
    }

    const database = path && path[0]

    if (!database) {
      throw new Error('[URI] missing: database!')
    }

    let ssl: undefined | string | SSLConfig

    if (params.ssl) {
      try {
        const sslParams: SSLParams = JSON.parse(params.ssl)

        ssl = {
          ciphers: sslParams.ciphers,
          passphrase: sslParams.passphrase,
          rejectUnauthorized: sslParams.rejectUnauthorized
        }

        if (sslParams.ca) {
          ssl.ca = readFileSync(resolvePath(sslParams.ca))
        }
        if (sslParams.cert) {
          ssl.cert = readFileSync(resolvePath(sslParams.cert))
        }
        if (sslParams.key) {
          ssl.key = readFileSync(resolvePath(sslParams.key))
        }
      } catch (_) {
        ssl = params.ssl
      }
    }

    const databaseConfig: MySQLDatabaseConfig = {
      database,
      host,
      port,
      user,
      password,
      multipleStatements: true,
      ssl
    }

    const engineConfig: MySQLEngineConfig = {
      migrationRecordTable: params.synor_migration_record_table
    }

    return {
      databaseConfig,
      engineConfig
    }
  } catch (error) {
    throw new SynorError('Invalid DatabaseURI', error)
  }
}
