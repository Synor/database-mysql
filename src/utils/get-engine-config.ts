import { SynorError } from '@synor/core'

type MySQLDatabaseEngineConfig = {
  migrationTableName: string
}

export function getEngineConfig(uri: string): MySQLDatabaseEngineConfig {
  const { protocol, searchParams: params } = new URL(uri)

  if (protocol !== 'mysql:') {
    throw new SynorError(`Invalid DatabaseURI`)
  }

  const migrationTableName =
    params.get('synor-migration-record-table') || 'synor_migration_record'

  return {
    migrationTableName
  }
}
