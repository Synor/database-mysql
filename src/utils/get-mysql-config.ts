import { SynorError } from '@synor/core'

type ConnectionConfig = import('mysql').ConnectionConfig

type MySQLDatabaseConfig = Pick<ConnectionConfig, 'ssl'> &
  Required<
    Pick<
      ConnectionConfig,
      'database' | 'host' | 'port' | 'user' | 'password' | 'multipleStatements'
    >
  >

export function getMySQLConfig(uri: string): MySQLDatabaseConfig {
  const {
    protocol,
    pathname,
    hostname,
    port,
    username,
    password,
    searchParams
  } = new URL(uri)

  if (protocol !== 'mysql:') {
    throw new SynorError(`Invalid DatabaseURI`)
  }

  let ssl: ConnectionConfig['ssl']

  if (searchParams.has('ssl')) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const sslRaw = searchParams.get('ssl')!

    try {
      ssl = JSON.parse(sslRaw)
    } catch {
      ssl = sslRaw
    }
  }

  return {
    database: pathname.replace(/^\//, ''),
    host: hostname,
    port: Number(port),
    user: username,
    password: password,
    multipleStatements: true,
    ssl
  }
}
