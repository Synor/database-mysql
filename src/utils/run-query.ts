type Connection = import('mysql').Connection

export type QueryValue = boolean | number | string | Date

export async function runQuery<T = any>(
  connection: Connection,
  query: string,
  values: QueryValue[] = []
): Promise<T> {
  return new Promise((resolve, reject) => {
    connection.query(query, values, (err, results) => {
      if (err) {
        reject(err)
        return
      }

      resolve(results)
    })
  })
}
