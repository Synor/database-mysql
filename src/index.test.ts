import { createConnection } from 'mysql'
import MySQLEngine, { MySQLDatabaseEngine } from './index'
import { getMySQLConfig } from './utils/get-mysql-config'
import { runQuery } from './utils/run-query'

type Connection = import('mysql').Connection

type GetAdvisoryLockId = import('@synor/core').GetAdvisoryLockId
type GetUserInfo = import('@synor/core').GetUserInfo
type MigrationSource = import('@synor/core').MigrationSource

jest.setTimeout(120 * 1000)

jest.mock('perf_hooks')

const getTableColumnCount = async (
  connection: Connection,
  tableName: string,
  databaseName: string
): Promise<number> =>
  runQuery(
    connection,
    `SELECT column_name FROM information_schema.columns WHERE table_name = ? AND table_schema = ?;`,
    [tableName, databaseName]
  ).then(rows => rows.length)

const migrationSource: Record<'01.do' | '01.undo', MigrationSource> = {
  '01.do': {
    version: '01',
    type: 'do',
    title: 'Test One',
    body: 'SELECT 1;',
    hash: 'hash-01-do'
  },
  '01.undo': {
    version: '01',
    type: 'undo',
    title: 'Test One',
    body: 'SELEC -1;',
    hash: 'hash-01-undo'
  }
}

const baseVersion = '0'
const getAdvisoryLockId: GetAdvisoryLockId = (databaseName, ...names) => {
  return [databaseName, names.join()]
}
const getUserInfo: GetUserInfo = () => Promise.resolve(`Jest`)

describe('module exports', () => {
  test('default export exists', () => {
    expect(typeof MySQLEngine).toBe('function')
  })

  test('named export exists', () => {
    expect(typeof MySQLDatabaseEngine).toBe('function')
  })

  test('default and named exports are same', () => {
    expect(MySQLEngine).toBe(MySQLDatabaseEngine)
  })
})

describe('initialization', () => {
  let uri: Parameters<typeof MySQLDatabaseEngine>[0]
  const helpers: Parameters<typeof MySQLDatabaseEngine>[1] = {
    baseVersion,
    getAdvisoryLockId,
    getUserInfo
  }

  beforeEach(() => {
    uri = 'mysql://root:root@127.0.0.1:3306/synor'
    helpers.baseVersion = baseVersion
    helpers.getAdvisoryLockId = getAdvisoryLockId
    helpers.getUserInfo = getUserInfo
  })

  test.each([undefined, null, 0])('throws if uri is %s', uri => {
    expect(() => MySQLDatabaseEngine(uri as any, helpers)).toThrow()
  })

  test('throws if uri is empty', () => {
    uri = ' '
    expect(() => MySQLDatabaseEngine(uri, helpers)).toThrow()
  })

  describe('helpers validation', () => {
    beforeEach(() => {
      helpers.getAdvisoryLockId = getAdvisoryLockId
      helpers.getUserInfo = getUserInfo
    })

    test(`throws if getAdvisoryLockId is missing`, () => {
      delete helpers.getAdvisoryLockId
      expect(() => MySQLDatabaseEngine(uri, helpers)).toThrow()
    })

    test(`throws if getAdvisoryLockId is not function`, () => {
      helpers.getAdvisoryLockId = '' as any
      expect(() => MySQLDatabaseEngine(uri, helpers)).toThrow()
      helpers.getAdvisoryLockId = null as any
      expect(() => MySQLDatabaseEngine(uri, helpers)).toThrow()
    })

    test(`throws if getUserInfo is missing`, () => {
      delete helpers.getUserInfo
      expect(() => MySQLDatabaseEngine(uri, helpers)).toThrow()
    })

    test(`throws if getUserInfo is not function`, () => {
      helpers.getUserInfo = '' as any
      expect(() => MySQLDatabaseEngine(uri, helpers)).toThrow()
      helpers.getUserInfo = null as any
      expect(() => MySQLDatabaseEngine(uri, helpers)).toThrow()
    })
  })
})

const databaseName = 'synor'
const tableName = 'test_record'
const params = `synor-migration-record-table=${tableName}`
const uri = `mysql://root:root@127.0.0.1:3306/${databaseName}?${params}`

describe('methods: {open,close}', () => {
  let connection: Connection

  let engine: ReturnType<typeof MySQLDatabaseEngine>

  beforeAll(async () => {
    connection = createConnection(getMySQLConfig(uri))
    connection.connect()

    await runQuery(connection, 'DROP TABLE IF EXISTS ??;', [tableName])
  })

  afterAll(() => {
    connection.end()
  })

  beforeEach(() => {
    engine = MySQLDatabaseEngine(uri, {
      baseVersion,
      getAdvisoryLockId,
      getUserInfo
    })
  })

  let recordColumnCount: number

  test('open (first run)', async () => {
    recordColumnCount = await getTableColumnCount(
      connection,
      tableName,
      'synor'
    )

    expect(recordColumnCount).toBe(0)

    await expect(engine.open()).resolves.toBeUndefined()

    recordColumnCount = await getTableColumnCount(
      connection,
      tableName,
      'synor'
    )

    expect(recordColumnCount).toBeGreaterThan(0)
  })

  test('open (after first run)', async () => {
    await expect(
      getTableColumnCount(connection, tableName, databaseName)
    ).resolves.toBe(recordColumnCount)

    await expect(engine.open()).resolves.toBeUndefined()

    await expect(
      getTableColumnCount(connection, tableName, databaseName)
    ).resolves.toBe(recordColumnCount)
  })

  test('close', async () => {
    await expect(engine.close()).resolves.toBeUndefined()
  })
})

describe('methods: {lock,unlock}', () => {
  test('can lock & unlock', async () => {
    const engine = MySQLDatabaseEngine(uri, {
      baseVersion,
      getAdvisoryLockId,
      getUserInfo
    })

    await engine.open()

    await expect(engine.lock()).resolves.toBeUndefined()

    await expect(engine.unlock()).resolves.toBeUndefined()

    await engine.close()
  })

  test('can not get multiple lock at once', async () => {
    const engineOne = MySQLDatabaseEngine(uri, {
      baseVersion,
      getAdvisoryLockId,
      getUserInfo
    })

    const engineTwo = MySQLDatabaseEngine(uri, {
      baseVersion,
      getAdvisoryLockId,
      getUserInfo
    })

    const callOrder: Array<'lock-1' | 'unlock-1' | 'lock-2' | 'unlock-2'> = []

    await Promise.all([engineOne.open(), engineTwo.open()])

    await engineOne.lock().then(() => {
      callOrder.push('lock-1')
    })

    await Promise.all([
      engineTwo.lock().then(() => {
        callOrder.push('lock-2')
      }),
      engineOne.unlock().then(() => {
        callOrder.push('unlock-1')
      })
    ])

    await engineTwo.unlock().then(() => {
      callOrder.push('unlock-2')
    })

    expect(callOrder).toEqual(['lock-1', 'unlock-1', 'lock-2', 'unlock-2'])

    await Promise.all([engineOne.close(), engineTwo.close()])
  })

  test('lock throws if failed to get', async () => {
    const queries = jest.requireActual('./queries')

    jest.spyOn(queries, 'getQueryStore').mockImplementationOnce((...args) => {
      const queryStore = queries.getQueryStore(...args)
      queryStore.getLock = () => Promise.resolve(0)
      return queryStore
    })

    const engine = MySQLDatabaseEngine(uri, {
      baseVersion,
      getAdvisoryLockId,
      getUserInfo
    })

    await engine.open()

    await expect(engine.lock()).rejects.toThrow()

    await engine.close()
  })

  test('unlock throws if not locked', async () => {
    const engine = MySQLDatabaseEngine(uri, {
      baseVersion,
      getAdvisoryLockId,
      getUserInfo
    })

    await engine.open()

    await expect(engine.unlock()).rejects.toThrow()

    await engine.close()
  })
})

describe('methods', () => {
  let connection: Connection

  let engine: ReturnType<typeof MySQLDatabaseEngine>

  const OriginalDate = Date

  beforeAll(async () => {
    global.Date = class extends OriginalDate {
      constructor() {
        super('2020-01-01T00:00:00.000Z')
      }
    } as typeof global.Date

    connection = createConnection(getMySQLConfig(uri))
    connection.connect()

    await runQuery(connection, 'DROP TABLE IF EXISTS ??;', [tableName])
  })

  afterAll(() => {
    connection.end()

    global.Date = OriginalDate
  })

  beforeEach(async () => {
    engine = MySQLDatabaseEngine(uri, {
      baseVersion,
      getAdvisoryLockId,
      getUserInfo
    })

    await engine.open()
  })

  afterEach(async () => {
    await engine.close()
  })

  test('drop', async () => {
    await expect(
      getTableColumnCount(connection, tableName, databaseName)
    ).resolves.toBeGreaterThan(0)
    await expect(engine.drop()).resolves.toBeUndefined()
    await expect(
      getTableColumnCount(connection, tableName, databaseName)
    ).resolves.toBe(0)
  })

  test('run', async () => {
    await expect(engine.run(migrationSource['01.do'])).resolves.toBeUndefined()
    await expect(engine.run(migrationSource['01.undo'])).rejects.toThrow()

    await expect(
      runQuery(connection, `SELECT * FROM ??;`, [tableName])
    ).resolves.toMatchSnapshot()

    await engine.drop()
  })

  test('repair', async () => {
    await expect(engine.run(migrationSource['01.do'])).resolves.toBeUndefined()
    await expect(engine.run(migrationSource['01.undo'])).rejects.toThrow()

    const [record] = await runQuery<any[]>(
      connection,
      `SELECT id FROM ?? WHERE version = ? AND type = ?;`,
      [
        tableName,
        migrationSource['01.do'].version,
        migrationSource['01.do'].type
      ]
    )

    await expect(
      engine.repair([
        { id: record.id, hash: `${migrationSource['01.do'].hash}-repaired` }
      ])
    ).resolves.toBeUndefined()

    await expect(
      runQuery(connection, `SELECT * FROM ??;`, [tableName])
    ).resolves.toMatchSnapshot()

    await engine.drop()
  })

  test('records', async () => {
    await expect(engine.run(migrationSource['01.do'])).resolves.toBeUndefined()
    await expect(engine.run(migrationSource['01.undo'])).rejects.toThrow()

    await expect(engine.records()).resolves.toMatchSnapshot()

    await engine.drop()
  })
})
