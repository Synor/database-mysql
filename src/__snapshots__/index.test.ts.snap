// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods records 1`] = `
Array [
  Object {
    "appliedAt": 2020-01-01T00:00:00.000Z,
    "appliedBy": "Synor",
    "dirty": false,
    "executionTime": 0,
    "hash": "",
    "id": 1,
    "title": "Base Migration",
    "type": "do",
    "version": "0",
  },
  Object {
    "appliedAt": 2020-01-01T00:00:00.000Z,
    "appliedBy": "Jest",
    "dirty": false,
    "executionTime": 0,
    "hash": "hash-01-do",
    "id": 2,
    "title": "Test One",
    "type": "do",
    "version": "01",
  },
  Object {
    "appliedAt": 2020-01-01T00:00:00.000Z,
    "appliedBy": "Jest",
    "dirty": true,
    "executionTime": 0,
    "hash": "hash-01-undo",
    "id": 3,
    "title": "Test One",
    "type": "undo",
    "version": "01",
  },
]
`;

exports[`methods repair 1`] = `
Array [
  RowDataPacket {
    "applied_at": 2020-01-01T00:00:00.000Z,
    "applied_by": "Synor",
    "dirty": 0,
    "execution_time": 0,
    "hash": "",
    "id": 1,
    "title": "Base Migration",
    "type": "do",
    "version": "0",
  },
  RowDataPacket {
    "applied_at": 2020-01-01T00:00:00.000Z,
    "applied_by": "Jest",
    "dirty": 0,
    "execution_time": 0,
    "hash": "hash-01-do-repaired",
    "id": 2,
    "title": "Test One",
    "type": "do",
    "version": "01",
  },
]
`;

exports[`methods run 1`] = `
Array [
  RowDataPacket {
    "applied_at": 2020-01-01T00:00:00.000Z,
    "applied_by": "Synor",
    "dirty": 0,
    "execution_time": 0,
    "hash": "",
    "id": 1,
    "title": "Base Migration",
    "type": "do",
    "version": "0",
  },
  RowDataPacket {
    "applied_at": 2020-01-01T00:00:00.000Z,
    "applied_by": "Jest",
    "dirty": 0,
    "execution_time": 0,
    "hash": "hash-01-do",
    "id": 2,
    "title": "Test One",
    "type": "do",
    "version": "01",
  },
  RowDataPacket {
    "applied_at": 2020-01-01T00:00:00.000Z,
    "applied_by": "Jest",
    "dirty": 1,
    "execution_time": 0,
    "hash": "hash-01-undo",
    "id": 3,
    "title": "Test One",
    "type": "undo",
    "version": "01",
  },
]
`;
